[RW-9408][risk=low] Create and delete RStudio GKE Apps

Peter-Lavigne opened the pull request:
Description:

Implements [RW-9408](https://precisionmedicineinitiative.atlassian.net/browse/RW-9408)

Allows users to create and delete RStudio GKE Apps.

This is toggled by the newly added `enableRStudioGKEApp` feature flag. I created two tickets ([1](https://precisionmedicineinitiative.atlassian.net/browse/RW-9645), [2](https://precisionmedicineinitiative.atlassian.net/browse/RW-9646)) to track enabling and removing this feature flag.

To validate this, you can open RStudio [using this method](https://pmi-engteam.slack.com/archives/D042SAZTEPJ/p1677596962109319). You can find the proxy URL in the response to the call to the apps endpoint when the sidebar opens.

risk=low because RStudio data exfil isn't implemented yet; if a user in prod somehow accessed this it would be a problem.

Before launching:
<img width="505" alt="image" src="https://user-images.githubusercontent.com/31020403/221980517-88843697-34a9-433a-bd14-69b7c1165940.png">

After launching:
<img width="510" alt="image" src="https://user-images.githubusercontent.com/31020403/221983171-3ee3834b-88b9-4ff8-a37e-631740092ba0.png">

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have added explanatory comments where the logic is not obvious
- [x] I have run and tested this change locally, and my testing process is described here
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [x] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [x] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert 


[RW-9408]: https://precisionmedicineinitiative.atlassian.net/browse/RW-9408?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Peter-Lavigne commented on ui/src/app/components/apps-panel/expanded-app.tsx:
```diff
@@ -147,6 +154,75 @@ const CromwellButtonRow = (props: {
   );
 };
 
+const RStudioButtonRow = (props: {
```
I considered combining this with the Cromwell one, but the number of small differences at the moment would make this pretty messy.

I expect them to converge over time. When I finish the [story to open RStudio](https://precisionmedicineinitiative.atlassian.net/browse/RW-9613) I'll reconsider combining them.

jmthibault79 replied:
makes sense

jmthibault79 reviewed (commented):
(not a full review - just a few things I wanted to look at)

I also have other thoughts about an overall feature flag strategy for User Apps that I'd like to chat about - at Eng Sync maybe?

Peter-Lavigne commented:
As we talked about earlier, I converted the `showAppsPanel` feature flag to a server-side `enableCromwellGKEApp`. We now calculate `showAppsPanel` by checking if either app is enabled.

In that commit, I also
- Added a server-side check for the flag when creating Cromwell apps
- Converted `DisplayIcon` to a React component so it could use hooks (a test was failing)

The app with only RStudio enabled:
<img width="518" alt="image" src="https://user-images.githubusercontent.com/31020403/222239448-0c3de1b8-949a-4b4a-bffb-df068d3ba6af.png">

Peter-Lavigne commented on api/src/main/java/org/pmiops/workbench/api/AppsController.java:
```diff
@@ -45,6 +45,10 @@ public ResponseEntity<EmptyResponse> createApp(
     validateCanPerformApiAction(dbWorkspace);
+    if (createAppRequest.getAppType() == AppType.RSTUDIO
+        && !configProvider.get().featureFlags.enableRStudioGKEApp) {
+      throw new UnsupportedOperationException("API not supported.");
```
As we talked about, we're leaving this since there are potential security concerns at the moment.

jmthibault79 replied:
Yep.  I'd also like to hear @yonghaoy 's opinion on this.

The logic is that we have not yet designated *all* potential apps as OK from a security perspective (exfil, etc) so we should be conservative: allow only those we have explicitly enabled in the environment.

Peter-Lavigne commented on api/src/main/resources/workbench-api.yaml:
```diff
@@ -4673,6 +4673,10 @@ definitions:
+      enableRStudioGKEApp:
+        type: boolean
+        default: false
+        description: Whether users are able to use the RStudio GKE app.
```
Since we're not removing the API check, I'm leaving this as-is.

Peter-Lavigne commented on api/src/main/java/org/pmiops/workbench/api/AppsController.java:
```diff
@@ -2,15 +2,12 @@
-import org.pmiops.workbench.model.ListAppsResponse;
-import org.pmiops.workbench.model.UserAppEnvironment;
-import org.pmiops.workbench.model.WorkspaceAccessLevel;
+import org.pmiops.workbench.model.*;
```
I've updated my IntelliJ to avoid this. The official instructions for this are interesting:

<img width="690" alt="image" src="https://user-images.githubusercontent.com/31020403/222240817-f1cf6931-db81-40ab-a8e1-ae2aa3498198.png">

jmthibault79 replied:
That is also my maximum tab count

jmthibault79 commented on api/src/test/java/org/pmiops/workbench/api/AppsControllerTest.java:
```diff
@@ -84,6 +84,7 @@ public void setUp() {
 
     user = new DbUser();
     createAppRequest = new CreateAppRequest().appType(AppType.RSTUDIO);
+    config.featureFlags.enableRStudioGKEApp = true;
```
I'd prefer to remove this and set them explicitly in the tests

Peter-Lavigne replied:
Good idea, done.

jmthibault79 commented on api/src/test/java/org/pmiops/workbench/api/AppsControllerTest.java:
```diff
@@ -109,6 +110,24 @@ public void testCreateAppFail_validateActiveBilling() {
         ForbiddenException.class, () -> controller.createApp(WORKSPACE_NS, createAppRequest));
   }
 
+  @Test
```
Could you add "enableRStudio = true" to this one, to show that it doesn't help, and the equivalent to the next one?

I'd also like to see success cases for each.

Peter-Lavigne replied:
Done 👍

jmthibault79 commented on ui/src/app/components/apps-panel.tsx:
```diff
@@ -58,6 +55,14 @@ export const AppsPanel = (props: {
+  // in display order
+  const appsToDisplay = [
+    UIAppType.JUPYTER,
+    ...(config.enableRStudioGKEApp ? [UIAppType.RSTUDIO] : []),
```
nice

jmthibault79 commented on ui/src/app/components/apps-panel/utils.tsx:
```diff
@@ -146,3 +160,7 @@ export const toUserEnvironmentStatusByAppType = (
     : fromUserAppStatus(status as AppStatus);
+
+export const showAppsPanel = (config: ConfigResponse) => {
+  return config.enableCromwellGKEApp || config.enableRStudioGKEApp;
```
I like it

jmthibault79 commented on ui/src/app/components/help-sidebar-icons.tsx:
```diff
@@ -668,7 +681,7 @@ export const HelpSidebarIcons = (props: HelpSidebarIconsProps) => {
               }}
             >
-              {displayIcon(icon, props)}
+              <DisplayIcon {...{ ...props, icon }} />
```
this change seems so obvious now that you've done it, heh

jmthibault79 commented on ui/src/app/components/apps-panel.spec.tsx:
```diff
@@ -64,11 +68,14 @@ describe('AppsPanel', () => {
+      config: { ...defaultServerConfig, enableCromwellGKEApp: true },
+    });
     registerApiClient(AppsApi, appsStub);
+    appsStub.createApp = () => Promise.resolve({});
```
is this needed?  the stub exists (added recently)

Peter-Lavigne replied:
I did this out of fear of the issue described [here](https://github.com/all-of-us/workbench/pull/7450#discussion_r1122321488). Adding workaround code probably isn't the right path forward, though. The tests still pass when I remove it, so I'll remove it.

jmthibault79 commented on ui/src/app/components/apps-panel.spec.tsx:
```diff
@@ -136,7 +143,7 @@ describe('AppsPanel', () => {
   );
 
   // these tests assume that there are no Jupyter runtimes
-  // so what these tests actually show is whether Cromwell is an ActiveApp
```
nit: the original text remains true here.  But this text is also true.  I dunno.

Peter-Lavigne replied:
Yeah, there's no great way to generalize a test for "all apps" since you need to choose one of them for the test. I think the test is reflective enough of all apps at the moment that I'm ok with it.

One alternative is to re-run these tests for each app type but that seems more complex than it's worth.

jmthibault79 commented on ui/src/app/components/apps-panel.spec.tsx:
```diff
@@ -168,10 +175,11 @@ describe('AppsPanel', () => {
   );
 
-  it('should render ActiveApps only, when both Jupyter and Cromwell are RUNNING', async () => {
+  it('should render ActiveApps only, when all apps are RUNNING', async () => {
```
cool

jmthibault79 commented on ui/src/app/components/apps-panel.spec.tsx:
```diff
@@ -264,4 +283,124 @@ describe('AppsPanel', () => {
+    launchButton().simulate('click');
+    await waitOneTickAndUpdate(wrapper);
+
+    expect(appsStub.createApp).toHaveBeenCalledWith(
```
this is a simpler syntax than I've been using.  thanks

jmthibault79 commented on ui/src/app/components/help-sidebar-icons.tsx:
```diff
@@ -628,7 +641,7 @@ export const HelpSidebarIcons = (props: HelpSidebarIconsProps) => {
   );
 
   if (
```
We need to continue to display `'apps'` for RStudio here, so I think we need 3 ifs now.  Let's move the access check to the outside...

```suggestion
  if (WorkspacePermissionsUtil.canWrite(workspace.accessLevel)) {
    if (showAppsPanel(config)) {
      keys.push('apps');
    }
    if (config.enableCromwellGKEApp) {
      keys.push('cromwellConfig');
    }
    keys.push('runtimeConfig', 'terminal');
  }
```

Peter-Lavigne replied:
I feel like I'm being immediately punished for [generalizing tests across apps](https://github.com/all-of-us/workbench/pull/7450#discussion_r1122337334)! The test `should display apps icon for writable workspaces when showAppsPanel is true` would have caught this if we stubbed `showAppsPanel` to be true instead of proxying for it via `enableCromwellGKEApp`.

I modified the test and added a new one that would have caught this.

Peter-Lavigne replied:
In general, I'm still not sure whether stubbing helper methods or exhausting their parameters is a better testing approach.

Peter-Lavigne commented on ui/src/app/components/apps-panel.spec.tsx:
```diff
@@ -64,11 +68,14 @@ describe('AppsPanel', () => {
+    appsStub.createApp = () => Promise.resolve({});
     registerApiClient(NotebooksApi, new NotebooksApiStub());
     registerApiClient(RuntimeApi, runtimeStub);
+    workspaceStub.billingStatus = BillingStatus.ACTIVE;
```
Unfortunately, I think it's still needed because we set `billingStatus` to `INACTIVE` in one of the tests. This causes the underlying `workspaceDataStub`'s `billingStatus` field to become `INACTIVE` for the remainder of the tests. I didn't find any fixes I liked after fiddling with it for a few minutes.

There's a larger problem here that many of our JS tests aren't independent of one another.

If you suggest a good fix for this I'll add it, otherwise I'll leave it as-is.

jmthibault79 reviewed (approved):
I am able to successfully create and delete RStudios locally.

There is one issue with the displaying of the Apps Panel however

Peter-Lavigne commented:
With an apps panel:

<img width="515" alt="Screen Shot 2023-03-01 at 5 06 38 PM" src="https://user-images.githubusercontent.com/31020403/222275846-104ae647-fe6c-4c26-a139-f6ce8a5795e3.png">
